### ISSUE TITLE
PivotGrid - Excel Export Enhancements

### ISSUE THREAD

**Issue by EugeniyKiyashko** [DevExpress/DevExtreme#13847]
## The Problem

The [current export implementation](https://js.devexpress.com/Documentation/ApiReference/UI_Widgets/dxPivotGrid/Configuration/export/) doesn't allow the PivotGrid to support the following use cases:

- Customize Excel cells
- Customize Excel workbooks and worksheets
- Export the Field Panel
- Add extra worksheets with custom data
- Use custom formats
- Indicate export progress
- Protect Excel cells and worksheets

## The Proposed Solution

We are considering the possibility of using [ExcelJS](https://github.com/exceljs/exceljs) instead of our own implementation. ExcelJS is a third-party library for reading, manipulating, and writing spreadsheet data and styles to Excel and JSON. We already use ExcelJS for [export in our DataGrid](https://js.devexpress.com/Demos/WidgetsGallery/Demo/DataGrid/ExcelJSOverview/).

## Implementation Details

We will add a static `exportPivotGrid` method that will accept an object with the following fields:

```js
component: PivotGrid, // A PivotGrid instance
worksheet: Object, // An Excel worksheet to which the PivotGrid should be exported
customizeCell: Function, // Customizes an Excel cell after creation
topLeftCell: Object | String, // A cell used as a start position for export
loadPanel: Object // Configures the load panel
```

The ExcelJS API will be used to create a worksheet and customize its cells.

### Customize Cell Appearance
The `customizeCell` function will allow you to apply [flexible customizations](https://github.com/exceljs/exceljs#styles) to individual [pivot grid cells](https://js.devexpress.com/Documentation/ApiReference/UI_Widgets/dxPivotGrid/Pivot_Grid_Cell/):

<p align="center">
  <img src="https://user-images.githubusercontent.com/57402891/90413022-95326900-e0b6-11ea-99c4-bd91449ea038.png">
</p>

```js
DevExpress.excelExporter.exportPivotGrid({
    ...
    customizeCell: function({excelCell, pivotCell}) {
        if(pivotCell.area === 'data') {
            excelCell.fill = { type: 'pattern', pattern:'solid', fgColor: { argb: "FFEB9C" } };
            excelCell.font = { color: { argb: "9C6500" }};
            excelCell.border = { style: 'thin', color: { argb: 'FF7E7E7E' } };
        }
    }
}
```

### Add Custom Header and Footer

You can use the ExcelJS API to add custom [images](https://github.com/exceljs/exceljs#images) and [text strings](https://github.com/exceljs/exceljs#rich-text-value) to any cell of the exported worksheet. This capability allows you to create a custom header and footer in the document:

<p align="center">
  <img src="https://user-images.githubusercontent.com/57402891/90400286-eb96ac00-e0a4-11ea-9e76-b1477a91a4b3.png">
</p>

```js
DevExpress.excelExporter.exportPivotGrid({
    ...
}).then(function(cellRange) {
	// Header
	var headerRow = worksheet.getRow(1);
	headerRow.height = 70; 
	worksheet.mergeCells('D1:G1');
	headerRow.getCell(4).value = 'Sales Amount by Region';
	headerRow.getCell(4).font = { name: 'Segoe UI Light', size: 22, bold: true };

	// Footer
	var footerRowIndex = cellRange.to.row + 2;
	var footerRow = worksheet.getRow(footerRowIndex);
	worksheet.mergeCells(footerRowIndex, 1, footerRowIndex, 8);

	footerRow.getCell(1).value = 'www.wikipedia.org';
	footerRow.getCell(1).font = { color: { argb: 'BFBFBF' }, italic: true };
	footerRow.getCell(1).alignment = { horizontal: 'right' };
})
```

### Apply a Custom Cell Format

You will be able to specify any format for any cell or even use conditional formatting:

<p align="center">
  <img src="https://user-images.githubusercontent.com/57402891/90411731-dfb2e600-e0b4-11ea-8ee7-ded58f706ff3.png">
</p>

```js
DevExpress.excelExporter.exportPivotGrid({
    component: e.component,
    worksheet: worksheet,
    customizeCell: function({excelCell, pivotCell}) {
        if(pivotCell.area === 'data') {
            if(pivotCell.dataType === 'number'){
                excelCell.numFmt = '$#,##0';
            } else {
                excelCell.numFmt = 'mm/dd/yyyy';
            }
        }
    }        
})
```

### Indicate Export Progress

Export progress can be indicated by a load panel:

<p align="center">
  <img src="https://user-images.githubusercontent.com/57402891/84234821-8733fb00-aafd-11ea-9574-7f03f4344f40.png">
</p>

```js
DevExpress.excelExporter.exportPivotGrid({
    component: e.component,
    worksheet: worksheet,
    loadPanel: {
        enabled: true,
        text: 'Export large data...'
    }
})
```

### Live Sandboxes

1. [Customize cell appearance](https://codepen.io/EugeniyKiyashko/pen/GRZjVEz)
1. [Custom header and footer](https://codepen.io/EugeniyKiyashko/pen/poybRwz)
1. [Apply a custom cell format](https://codepen.io/EugeniyKiyashko/pen/NWNrdJK)
1. [Export a chart](https://codepen.io/EugeniyKiyashko/pen/bGEQEBO)
1. [Indicate export progress](https://codepen.io/EugeniyKiyashko/pen/yLeNVNx)
1. [Export to CSV](https://codepen.io/EugeniyKiyashko/pen/xxZGREK)
1. [Customize cell borders according to the PivotGrid border settings](https://codepen.io/EugeniyKiyashko/pen/pogJEqa)
1. [Resize columns](https://codepen.io/EugeniyKiyashko/pen/LYGYzwQ)

### Take a Quick Poll
[Do you find the enhanced export in PivotGrid useful for your project?](https://docs.google.com/forms/d/17nP7HiGe5ILj1mK7Tjn6vojNJMIGUDdufdeDh6K547g/viewform?usp=sf_link)

### Get Notified of Updates

Subscribe to this thread or to our [Facebook](https://www.facebook.com/DevExpress.DevExtreme/) and [Twitter](https://twitter.com/devextreme) accounts for updates on this topic.

**Comment by jsdmitry** [comment]
The [Export to Excel API](https://js.devexpress.com/Demos/WidgetsGallery/Demo/PivotGrid/ExcelJSOverview/jQuery/Light/) feature is now available in the [v20.2 release](https://github.com/DevExpress/DevExtreme/releases/tag/20.2.3). I'm closing this thread. In case of bugs or questions, feel free to create new [GitHub issues](https://github.com/DevExpress/DevExtreme/issues) or tickets in our [Support Center](https://www.devexpress.com/Support/Center/).

**Comment by Lonli-Lokli** [comment]
Is exceljs a required dependency or it can be replaced with any other implementation?
